### PR TITLE
Add ICP support.

### PR DESCRIPTION
**Pull request by RmbRT**
- Used ic_ledger_types crate to interact with the ICP ledger.
- Added real deposits and payout. Kept the mocked interactions so that
  the walk-through tests still work.

### CODE REVIEW

File: src/error.rs [hunk 1]
```diff
@@ -46,6 +46,8 @@ pub enum Error {
 	/// When a state that is registered for dispute is older than the previously
 	/// registered state.
 	OutdatedState,
+	/// Error while interaction with the ledger.
```
Comment (matthiasgeihs) [inline src/error.rs:49]:
```suggestion
	/// Error while interacting with the ledger.
```

File: src/icp.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+pub struct Receiver<Q: TXQuerier> {
+	tx_querier: Q,
+	my_account: AccountIdentifier,
+	known_txs: BTreeSet<BlockHeight>, // set of block heights
+	unspent: BTreeMap<Memo, Amount>,  // received tokens per memo
+}
+
+#[async_trait]
+pub trait TXQuerier {
+	async fn query_tx(&self, block_height: u64) -> Option<TransactionNotification>;
```
Comment (matthiasgeihs) [inline src/icp.rs:39]:
```suggestion
	async fn query_tx(&self, block_height: BlockHeight) -> Option<TransactionNotification>;
```

File: src/icp.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+#[derive(Default)]
+pub struct MockTXQuerier {
+	txs: BTreeMap<u64, TransactionNotification>,
```
Comment (matthiasgeihs) [inline src/icp.rs:44]: replace `u64` with descriptive type?

File: src/icp.rs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+#[async_trait]
+impl TXQuerier for MockTXQuerier {
+	async fn query_tx(&self, block_height: u64) -> Option<TransactionNotification> {
```
Comment (matthiasgeihs) [inline src/icp.rs:49]:
```suggestion
	async fn query_tx(&self, block_height: BlockHeight) -> Option<TransactionNotification> {
```

File: src/icp.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	Transaction, DEFAULT_SUBACCOUNT,
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub const MAINNET_ICP_LEDGER: &str = "ryjl3-tyaaa-aaaaa-aaaba-cai";
+
+pub type Memo = u64;
+pub type BlockHeight = u64;
+
+pub struct Receiver<Q: TXQuerier> {
```
Comment (matthiasgeihs) [inline src/icp.rs:30]: Add documentation.

File: src/icp.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+	tx_querier: Q,
+	my_account: AccountIdentifier,
+	known_txs: BTreeSet<BlockHeight>, // set of block heights
+	unspent: BTreeMap<Memo, Amount>,  // received tokens per memo
+}
+
+#[async_trait]
+pub trait TXQuerier {
```
Comment (matthiasgeihs) [inline src/icp.rs:38]: Add documentation.

File: src/lib.rs [hunk 7]
```diff
@@ -44,13 +57,20 @@ pub struct CanisterState {
...
 #[ic_cdk_macros::update]
-/// Deposits funds for the specified participant into the specified channel.
-/// Please do NOT over-fund or fund channels that are already fully funded, as
-/// this can lead to a permanent LOSS OF FUNDS.
-fn deposit(funding: Funding, amount: Amount) -> Option<Error> {
-	STATE
-		.with(|s| s.borrow_mut().deposit(funding, amount))
-		.err()
+/// The user needs to call this with his transaction.
+async fn transaction_notification(block_height: u64) {
```
Comment (matthiasgeihs) [inline src/lib.rs:61]:
```suggestion
async fn process_funding_tx(block_height: u64) {
```
?

Comment (matthiasgeihs) [reply src/lib.rs:61]: Combine with `deposit`?

File: src/lib.rs [hunk 8]
```diff
@@ -44,13 +57,20 @@ pub struct CanisterState {
...
+	STATE.write().unwrap().process_icp_tx(block_height).await;
+}
+
+#[ic_cdk_macros::update]
+fn deposit(funding: Funding) -> Option<Error> {
```
Comment (matthiasgeihs) [inline src/lib.rs:66]: Add documentation.

File: src/lib.rs [hunk 9]
```diff
@@ -69,38 +91,114 @@ fn dispute(params: Params, state: FullySignedState) -> Option<Error> {
...
 	STATE
-		.with(|s| s.borrow_mut().conclude(params, state, blocktime()))
+		.write()
+		.unwrap()
+		.conclude(params, state, blocktime())
 		.err()
 }
 
 #[ic_cdk_macros::update]
 /// Withdraws the specified participant's funds from a settled channel.
```
Comment (matthiasgeihs) [inline src/lib.rs:101]: Why call it `settled` channel but the method is named `conclude`?

File: src/lib.rs [hunk 10]
```diff
@@ -69,38 +91,114 @@ fn dispute(params: Params, state: FullySignedState) -> Option<Error> {
...
+async fn withdraw(
+	request: WithdrawalRequest,
+	auth: L2Signature,
+) -> (Option<icp::BlockHeight>, Option<Error>) {
+	let result = withdraw_impl(request, auth).await;
+	(result.as_ref().ok().cloned(), result.err())
+}
+
+#[ic_cdk_macros::update]
+/// Withdraws the specified participant's funds from a settled channel.
```
Comment (matthiasgeihs) [inline src/lib.rs:111]:
```suggestion
/// Withdraws the specified participant's funds from a concluded channel.
```

File: src/lib.rs [hunk 11]
```diff
@@ -69,38 +91,114 @@ fn dispute(params: Params, state: FullySignedState) -> Option<Error> {
...
 
 #[ic_cdk_macros::update]
 /// Withdraws the specified participant's funds from a settled channel.
-fn withdraw(request: WithdrawalRequest, auth: L2Signature) -> (Option<Amount>, Option<Error>) {
-	let result = STATE.with(|s| s.borrow_mut().withdraw(request, auth, blocktime()));
+async fn withdraw(
+	request: WithdrawalRequest,
+	auth: L2Signature,
+) -> (Option<icp::BlockHeight>, Option<Error>) {
+	let result = withdraw_impl(request, auth).await;
```
Comment (matthiasgeihs) [inline src/lib.rs:106]: Why separate `withdraw_impl` from `withdraw`?

File: src/lib.rs [hunk 12]
```diff
@@ -69,38 +91,114 @@ fn dispute(params: Params, state: FullySignedState) -> Option<Error> {
...
+/// Withdraws the specified participant's funds from a settled channel.
+async fn withdraw_mocked(
+	request: WithdrawalRequest,
+	auth: L2Signature,
+) -> (Option<Amount>, Option<Error>) {
+	let result = STATE.write().unwrap().withdraw(request, auth, blocktime());
 	(result.as_ref().ok().cloned(), result.err())
 }
 
+async fn withdraw_impl(request: WithdrawalRequest, auth: L2Signature) -> Result<icp::BlockHeight> {
```
Comment (matthiasgeihs) [inline src/lib.rs:120]: Add documentation.

File: src/icp.rs [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+		}
+
+		if let Some(tx) = self.tx_querier.query_tx(block_height).await {
+			if !self.known_txs.insert(block_height) {
+				return false;
+			}
+			if tx.to != self.my_account {
+				return false;
+			}
+			*self.unspent.entry(tx.memo).or_insert(0.into()) += tx.get_amount();
```
Comment (matthiasgeihs) [inline src/icp.rs:122]: Where do you verify that the `tx.to` account is consistent with `tx.memo`?

File: src/lib.rs (hunk 11 above)
Comment (RmbRT) [reply src/lib.rs:106]: because of the `?` operator which cannot be used in non-`Result` functions.

File: src/icp.rs (hunk 13 above)
Comment (RmbRT) [reply src/icp.rs:122]: `tx.to` is the perun canister's account, and `tx.memo` is the first 8 bytes of the hash of the channel funding ID. They are completely unrelated.

File: src/lib.rs (hunk 7 above)
Comment (RmbRT) [reply src/lib.rs:61]: Currently I need it separately for easier testing.

File: src/icp.rs (hunk 13 above)
Comment (matthiasgeihs) [reply src/icp.rs:122]: OK, and who sets `tx.memo` initially? Is this done off-chain when forming the tx?

**Pull request merged**